lib/irb/ruby-lex.rb

```diff
@@ -8,489 +8,500 @@
+
+      case start_token&.event
+      when :on_tstring_beg
+        case start_token&.tok
+        when ?"      then ?"
+        when /^%.$/  then ?"
+        when /^%Q.$/ then ?"
+        when ?'      then ?'
+        when /^%q.$/ then ?'
         end
-      elsif t.tok.include?("\n")
-        past_first_newline = true
-        false
+      when :on_regexp_beg   then ?/
+      when :on_symbeg       then ?:
+      when :on_backtick     then ?`
+      when :on_qwords_beg   then ?]
+      when :on_words_beg    then ?]
+      when :on_qsymbols_beg then ?]
+      when :on_symbols_beg  then ?]
+      when :on_heredoc_beg
+        start_token&.tok =~ /<<[-~]?(['"`])\w+\1/
+        $1 || ?"
       else
-        false
+        nil
       end
     end
 
-    if index
-      first_token = nil
-      last_line_tokens = tokens[(index + 1)..(tokens.size - 1)]
-      last_line_tokens.each do |t|
-        unless [:on_sp, :on_ignored_sp, :on_comment].include?(t.event)
-          first_token = t
-          break
+    def check_termination_in_prev_line(code)
+      tokens = self.class.ripper_lex_without_warning(code, context: @context)
+      past_first_newline = false
+      index = tokens.rindex do |t|
+        # traverse first token before last line
+        if past_first_newline
+          if t.tok.include?("\n")
+            true
+          end
+        elsif t.tok.include?("\n")
+          past_first_newline = true
+          false
+        else
+          false
         end
       end
 
-      if first_token && first_token.state != Ripper::EXPR_DOT
-        tokens_without_last_line = tokens[0..index]
-        code_without_last_line = tokens_without_last_line.map(&:tok).join
-        opens_without_last_line = IRB::NestingParser.open_tokens(tokens_without_last_line)
-        if code_terminated?(code_without_last_line, tokens_without_last_line, opens_without_last_line)
-          return last_line_tokens.map(&:tok).join
+      if index
+        first_token = nil
+        last_line_tokens = tokens[(index + 1)..(tokens.size - 1)]
+        last_line_tokens.each do |t|
+          unless [:on_sp, :on_ignored_sp, :on_comment].include?(t.event)
+            first_token = t
+            break
+          end
+        end
+
+        if first_token && first_token.state != Ripper::EXPR_DOT
+          tokens_without_last_line = tokens[0..index]
+          code_without_last_line = tokens_without_last_line.map(&:tok).join
+          opens_without_last_line = NestingParser.open_tokens(tokens_without_last_line)
+          if code_terminated?(code_without_last_line, tokens_without_last_line, opens_without_last_line)
+            return last_line_tokens.map(&:tok).join
+          end
         end
       end
+      false
     end
-    false
+  end
+  # :startdoc:
+end
+
+class RubyLex < IRB::RubyLex
+  def initialize(_)
+    warn <<~MSG
+      RubyLex is a private component of IRB and referencing to it from outside of IRB is deprecated.
+    MSG
```

Comment:
What do you think about deprecate_constant?
```ruby
RubyLex = IRB::RubyLex
Object.deprecate_constant :RubyLex
```
It only warns with `ruby -w`. 
It can also warn use of `RubyLex::Const` and `RubyLex.class_method` but can't customize deprecation message.

Reply:
Hmm I did try that but when I referenced it an exception was raised. But I guess it's probably something wrong with my testing script. Have updated it now 👍